[WebXR Hit Test] Implement requestHitTestSource() method of XRSession interface

fujii opened the pull request:
#### 2682af21e790848860d4dd8c6237b4bf89cc44a2
<pre>
[WebXR Hit Test] Implement requestHitTestSource() method of XRSession interface
<a href="https://bugs.webkit.org/show_bug.cgi?id=301495">https://bugs.webkit.org/show_bug.cgi?id=301495</a>

Reviewed by Sergio Villar Senin.

Implemented a part of requestHitTestSource() method of XRSession interface. The
platform specific part and FakeXRDevice part aren&apos;t implemented yet.

Unskipped WPT WebXR tests with the current test results. The wpt tests were
failing due to a missing required `supportsImmersive` member of
FakeXRDeviceInit. Removed `required` keyword of `supportsImmersive` member of
FakeXRDeviceInit interface because it is deprecated.

Tests: imported/w3c/web-platform-tests/webxr/hit-test

* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_source_cancel.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_inputSources.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_refSpaces.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_states_regular.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_states_transient.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_transientInputSources.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_unlocalizable.https-expected.txt: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webxr/WebXRFrame.cpp:
(WebCore::WebXRFrame::getHitTestResults):
(WebCore::WebXRFrame::getHitTestResultsForTransientInput):
* Source/WebCore/Modules/webxr/WebXRFrame.h:
(WebCore::WebXRFrame::session const):
* Source/WebCore/Modules/webxr/WebXRHitTestResult.cpp:
(WebCore::WebXRHitTestResult::create):
(WebCore::WebXRHitTestResult::WebXRHitTestResult):
(WebCore::WebXRHitTestResult::getPose):
* Source/WebCore/Modules/webxr/WebXRHitTestResult.h:
* Source/WebCore/Modules/webxr/WebXRHitTestResult.idl:
* Source/WebCore/Modules/webxr/WebXRHitTestSource.cpp:
(WebCore::WebXRHitTestSource::create):
(WebCore::WebXRHitTestSource::WebXRHitTestSource):
(WebCore::WebXRHitTestSource::~WebXRHitTestSource):
(WebCore::WebXRHitTestSource::cancel):
* Source/WebCore/Modules/webxr/WebXRHitTestSource.h:
(WebCore::WebXRHitTestSource::handle const):
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::requestHitTestSource):
* Source/WebCore/Modules/webxr/WebXRSystem.h:
* Source/WebCore/Modules/webxr/XRHitTestTrackableType.h:
* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::frameTimerFired):
(WebCore::SimulatedXRDevice::requestHitTestSource):
(WebCore::SimulatedXRDevice::deleteHitTestSource):
* Source/WebCore/testing/WebFakeXRDevice.h:
* Source/WebCore/testing/WebXRTest.idl:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/Shared/XR/XRDeviceProxy.cpp:
(WebKit::XRDeviceProxy::requestHitTestSource):
(WebKit::XRDeviceProxy::deleteHitTestSource):
* Source/WebKit/Shared/XR/XRDeviceProxy.h:
* Source/WebKit/UIProcess/XR/PlatformXRCoordinator.h:
(WebKit::PlatformXRCoordinator::requestHitTestSource):
(WebKit::PlatformXRCoordinator::deleteHitTestSource):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::requestHitTestSource):
(WebKit::PlatformXRSystem::deleteHitTestSource):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.messages.in:
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::requestHitTestSource):
(WebKit::OpenXRCoordinator::deleteHitTestSource):
(WebKit::OpenXRCoordinator::populateFrameData):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.cpp:
(WebKit::PlatformXRSystemProxy::requestHitTestSource):
(WebKit::PlatformXRSystemProxy::deleteHitTestSource):
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.h:

Canonical link: <a href="https://commits.webkit.org/302646@main">https://commits.webkit.org/302646@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c22143e85e5aac526ddfac187cede505edebf78

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129718 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1979 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40574 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137107 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81182 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131589 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1928 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1868 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98818 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66648 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132665 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1474 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116181 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79496 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/129068 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1391 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34315 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80379 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109875 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34817 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139589 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1773 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1661 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107328 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1818 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112531 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107202 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1435 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31026 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54523 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20247 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1846 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65215 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1660 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1695 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1767 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

svillar commented on Source/WebCore/Modules/webxr/WebXRFrame.cpp:
```diff
@@ -388,14 +389,37 @@ ExceptionOr<bool> WebXRFrame::fillPoses(const Document& document, const Vector<R
+ExceptionOr<Vector<Ref<WebXRHitTestResult>>> WebXRFrame::getHitTestResults(const WebXRHitTestSource& source)
 {
-    return { };
+    // 1. If frame’s active boolean is false, throw an InvalidStateError and abort these steps.
```
I did a similar thing when implementing WebXR, i.e., copying the spec steps as comments. But I am no longer sure that's a good idea. Specs change and these comments become obsolete as any other comment. I think we should stick to WebKit's coding style and add them only if it's really a must to understand the code.

fujii replied:
Removed.

svillar commented on Source/WebCore/Modules/webxr/WebXRHitTestResult.cpp:
```diff
@@ -37,11 +37,31 @@ namespace WebCore {
+    auto exceptionOrPose = m_frame->populatePose(document, space, baseSpace);
+    if (exceptionOrPose.hasException())
+        return exceptionOrPose.releaseException();
+    auto nullOrPopulatedPose = exceptionOrPose.releaseReturnValue();
```
It's really an optional so I think it's better to call this just populatedPose

fujii replied:
Fixed.

svillar commented on Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
```diff
@@ -271,6 +272,41 @@ void PlatformXRSystem::submitFrame(IPC::Connection& connection)
+    auto* xrCoordinator = PlatformXRSystem::xrCoordinator();
+    if (!xrCoordinator)
+        return;
+    xrCoordinator->deleteHitTestSource(*page, source);
```
What about
if (auto* xrCoordinator = ...)
   xrCoordinator->deleteHit

fujii replied:
Fixed.

svillar reviewed (commented):
Looking good. I have just a few comments.

BTW the API tests are failing

svillar commented on Source/WebCore/Modules/webxr/WebXRHitTestResult.cpp:
```diff
@@ -37,11 +37,31 @@ namespace WebCore {
 WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(WebXRHitTestResult);
 
+Ref<WebXRHitTestResult> WebXRHitTestResult::create(WebXRFrame& frame, const PlatformXR::FrameData::HitTestResult& result)
+{
```
Is the `result` guaranteed to survive the WebXRHitTestResult? I'm worried about storing a reference here if that precondition is not met.

fujii replied:
Not storing a reference here. Copying a vector.

fujii replied:
Not a vector.  PlatformXR::FrameData::HitTestResult is a struct. Copying a struct.

svillar commented on Source/WebCore/Modules/webxr/WebXRHitTestSource.h:
```diff
@@ -27,20 +27,29 @@
     WTF_MAKE_TZONE_OR_ISO_ALLOCATED(WebXRHitTestSource);
 public:
-    static Ref<WebXRHitTestSource> create();
+    static Ref<WebXRHitTestSource> create(WebXRSession&, PlatformXR::HitTestSource);
```
I know the source is an unsigned right now but I wonder whether it would be still safer to use && here just in case it changes in the future.

fujii replied:
Fixed.

svillar commented on Source/WebCore/Modules/webxr/WebXRHitTestSource.cpp:
```diff
@@ -35,17 +36,35 @@ namespace WebCore {
-WebXRHitTestSource::WebXRHitTestSource() = default;
+WebXRHitTestSource::WebXRHitTestSource(WebXRSession& session, PlatformXR::HitTestSource&& source)
+    : m_session(WTFMove(session))
+    , m_source(source)
```
WTFMove

fujii replied:
Oops. Good catch. Will fix. Thank you.

fujii committed:
[WebXR Hit Test] Implement requestHitTestSource() method of XRSession interface
https://bugs.webkit.org/show_bug.cgi?id=301495

Reviewed by Sergio Villar Senin.

Implemented a part of requestHitTestSource() method of XRSession interface. The
platform specific part and FakeXRDevice part aren't implemented yet.

Unskipped WPT WebXR tests with the current test results. The wpt tests were
failing due to a missing required `supportsImmersive` member of
FakeXRDeviceInit. Removed `required` keyword of `supportsImmersive` member of
FakeXRDeviceInit interface because it is deprecated.

Tests: imported/w3c/web-platform-tests/webxr/hit-test

* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_source_cancel.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_inputSources.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_refSpaces.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_states_regular.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_states_transient.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_transientInputSources.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_unlocalizable.https-expected.txt: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webxr/WebXRFrame.cpp:
(WebCore::WebXRFrame::getHitTestResults):
(WebCore::WebXRFrame::getHitTestResultsForTransientInput):
* Source/WebCore/Modules/webxr/WebXRFrame.h:
(WebCore::WebXRFrame::session const):
* Source/WebCore/Modules/webxr/WebXRHitTestResult.cpp:
(WebCore::WebXRHitTestResult::create):
(WebCore::WebXRHitTestResult::WebXRHitTestResult):
(WebCore::WebXRHitTestResult::getPose):
* Source/WebCore/Modules/webxr/WebXRHitTestResult.h:
* Source/WebCore/Modules/webxr/WebXRHitTestResult.idl:
* Source/WebCore/Modules/webxr/WebXRHitTestSource.cpp:
(WebCore::WebXRHitTestSource::create):
(WebCore::WebXRHitTestSource::WebXRHitTestSource):
(WebCore::WebXRHitTestSource::~WebXRHitTestSource):
(WebCore::WebXRHitTestSource::cancel):
* Source/WebCore/Modules/webxr/WebXRHitTestSource.h:
(WebCore::WebXRHitTestSource::handle const):
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::requestHitTestSource):
* Source/WebCore/Modules/webxr/WebXRSystem.h:
* Source/WebCore/Modules/webxr/XRHitTestTrackableType.h:
* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::frameTimerFired):
(WebCore::SimulatedXRDevice::requestHitTestSource):
(WebCore::SimulatedXRDevice::deleteHitTestSource):
* Source/WebCore/testing/WebFakeXRDevice.h:
* Source/WebCore/testing/WebXRTest.idl:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/Shared/XR/XRDeviceProxy.cpp:
(WebKit::XRDeviceProxy::requestHitTestSource):
(WebKit::XRDeviceProxy::deleteHitTestSource):
* Source/WebKit/Shared/XR/XRDeviceProxy.h:
* Source/WebKit/UIProcess/XR/PlatformXRCoordinator.h:
(WebKit::PlatformXRCoordinator::requestHitTestSource):
(WebKit::PlatformXRCoordinator::deleteHitTestSource):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::requestHitTestSource):
(WebKit::PlatformXRSystem::deleteHitTestSource):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.messages.in:
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::requestHitTestSource):
(WebKit::OpenXRCoordinator::deleteHitTestSource):
(WebKit::OpenXRCoordinator::populateFrameData):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.cpp:
(WebKit::PlatformXRSystemProxy::requestHitTestSource):
(WebKit::PlatformXRSystemProxy::deleteHitTestSource):
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.h:

Canonical link: https://commits.webkit.org/302646@main
